btf_interp/src/lib.rs

```diff
@@ -2,50 +2,199 @@
     adjust_tape: bool,
     /// The pointer to the current element of tape.
-    pointer: usize,
+    head: usize,
+    program: &'a BrainFuckProgram,
 }
 
-impl<T> VirtualMachine<T> {
+impl<'a, T> VirtualMachine<'a, T> {
     /// Create VM based on the size, by default is 30 000. Also, it can be adjusted, by default it doesn't.
-    pub fn new(size: Option<NonZeroUsize>, adjust_tape: Option<bool>) -> Self {
+    pub fn new(
+        program: &'a BrainFuckProgram,
+        size: Option<NonZeroUsize>,
+        adjust_tape: Option<bool>,
+    ) -> Self {
         VirtualMachine {
             tape: Vec::new(),
             tape_size: size.map(NonZeroUsize::get).unwrap_or(3000),
             adjust_tape: adjust_tape.unwrap_or(false),
-            pointer: 0,
+            head: 0,
+            program,
         }
     }
 
-    /// Add element into tape. If tape size exceeded, element is not going to be added, error message is shown.
-    pub fn push(&mut self, item: T) -> Result<(), String> {
-        if self.tape.len() >= self.tape_size {
-            if !self.adjust_tape {
-                return Err(format!("Max tape size of {} reached.", self.tape_size));
-            }
-            self.tape_size += 1;
-        }
-        self.tape.push(item);
-        Ok(())
-    }
-
     /// Interpreter BF program into human-readable format.
-    pub fn interpreter(&self, bf_program: &BrainFuckProgram) {
-        for instruction_position in bf_program.instructions() {
+    pub fn interpreter(&self) {
+        for instruction_position in self.program.instructions() {
             println!(
                 "[{}:{}",
-                bf_program.filename().display(),
+                self.program.filename().display(),
                 instruction_position
             );
         }
     }
+
+    /// Go to the next element in tape. If tape size exceeded, error message is shown.
+    pub fn next_element(&mut self) -> Result<(), VMError> {
+        if self.head == self.tape_size - 1 {
+            return Err(VMError::NextElementNotReachable);
+        }
+        self.head += 1;
+        Ok(())
+    }
+
+    /// Go to the previous element in tape. If it is the first element, error message is shown.
+    pub fn previous_element(&mut self) -> Result<(), VMError> {
+        if self.head == 0 {
+            return Err(VMError::PreviousElementNotReachanble);
+        }
+        self.head -= 1;
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::BrainFuckProgram;
+    use crate::NonZeroUsize;
+    use crate::VMError;
+    use crate::VirtualMachine;
+
+    use std::fs::File;
+    use tempdir::TempDir;
+
+    #[test]
+    fn test_default_new_vm() {
+        let tmp_dir = TempDir::new("example").unwrap();
+        let file_path = tmp_dir.path().join("my-temporary-note.txt");
+        let tmp_file = File::create(&file_path).unwrap();
+
+        let program = BrainFuckProgram::from_file(&file_path).unwrap();
```

Comment:
Rather than *all* your tests needing a tempdir, do you not have a `from_string()` or similar?

Reply:
But I have to have path and file following the path